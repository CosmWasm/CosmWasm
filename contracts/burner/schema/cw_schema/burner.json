{
  "contract_name": "burner",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "InstantiateMsg",
        "description": "A placeholder where we don't take any input",
        "type": "struct",
        "properties": {}
      }
    ]
  },
  "execute": null,
  "query": null,
  "migrate": {
    "type": "v1",
    "root": 2,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      },
      {
        "name": "MigrateMsg",
        "type": "struct",
        "properties": {
          "delete": {
            "description": "Optional amount of items to delete in this call.\nIf it is not provided, nothing will be deleted.\nYou can delete further items in a subsequent execute call.",
            "value": 1
          },
          "payout": {
            "description": "The address we send all remaining balance to",
            "value": 0
          }
        }
      }
    ]
  },
  "sudo": null,
  "responses": null
}
