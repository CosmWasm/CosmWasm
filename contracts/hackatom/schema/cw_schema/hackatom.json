{
  "contract_name": "hackatom",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 1,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "InstantiateMsg",
        "type": "struct",
        "properties": {
          "beneficiary": {
            "value": 0
          },
          "verifier": {
            "value": 0
          }
        }
      }
    ]
  },
  "execute": {
    "type": "v1",
    "root": 1,
    "definitions": [
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      },
      {
        "name": "ExecuteMsg",
        "type": "enum",
        "cases": {
          "allocate_large_memory": {
            "description": "Allocate large amounts of memory without consuming much gas",
            "type": "named",
            "properties": {
              "pages": {
                "value": 0
              }
            }
          },
          "cpu_loop": {
            "description": "Infinite loop to burn cpu cycles (only run when metering is enabled)",
            "type": "named",
            "properties": {}
          },
          "memory_loop": {
            "description": "Infinite loop reading and writing memory",
            "type": "named",
            "properties": {}
          },
          "message_loop": {
            "description": "Infinite loop sending message to itself",
            "type": "named",
            "properties": {}
          },
          "panic": {
            "description": "Trigger a panic to ensure framework handles gracefully",
            "type": "named",
            "properties": {}
          },
          "release": {
            "description": "Releasing all funds in the contract to the beneficiary. This is the only \"proper\" action of this demo contract.",
            "type": "named",
            "properties": {}
          },
          "storage_loop": {
            "description": "Infinite loop making storage calls (to test when their limit hits)",
            "type": "named",
            "properties": {}
          },
          "user_errors_in_api_calls": {
            "description": "Starting with CosmWasm 0.10, some API calls return user errors back to the contract.\nThis triggers such user errors, ensuring the transaction does not fail in the backend.",
            "type": "named",
            "properties": {}
          }
        }
      }
    ]
  },
  "query": {
    "type": "v1",
    "root": 2,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      },
      {
        "name": "QueryMsg",
        "type": "enum",
        "cases": {
          "get_int": {
            "description": "GetInt returns a hardcoded u32 value",
            "type": "named",
            "properties": {}
          },
          "other_balance": {
            "description": "This returns cosmwasm_std::AllBalanceResponse to demo use of the querier",
            "type": "named",
            "properties": {
              "address": {
                "value": 0
              }
            }
          },
          "recurse": {
            "description": "Recurse will execute a query into itself up to depth-times and return\nEach step of the recursion may perform some extra work to test gas metering\n(`work` rounds of sha256 on contract).\nNow that we have Env, we can auto-calculate the address to recurse into",
            "type": "named",
            "properties": {
              "depth": {
                "value": 1
              },
              "work": {
                "value": 1
              }
            }
          },
          "verifier": {
            "description": "returns a human-readable representation of the verifier\nuse to ensure query path works in integration tests",
            "type": "named",
            "properties": {}
          }
        }
      }
    ]
  },
  "migrate": {
    "type": "v1",
    "root": 1,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "MigrateMsg",
        "description": "MigrateMsg allows a privileged contract administrator to run\na migration on the contract. In this (demo) case it is just migrating\nfrom one hackatom code to the same code, but taking advantage of the\nmigration step to set a new validator.\n\nNote that the contract doesn't enforce permissions here, this is done\nby blockchain logic (in the future by blockchain governance)",
        "type": "struct",
        "properties": {
          "verifier": {
            "value": 0
          }
        }
      }
    ]
  },
  "sudo": {
    "type": "v1",
    "root": 4,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "Uint128",
        "description": "A thin wrapper around u128 that is using strings for JSON encoding/decoding,\nsuch that the full u128 range can be used for clients that convert JSON numbers to floats,\nlike JavaScript and jq.\n\n# Examples\n\nUse `from` to create instances of this and `u128` to get the value out:\n\n```\n# use cosmwasm_std::Uint128;\nlet a = Uint128::from(123u128);\nassert_eq!(a.u128(), 123);\n\nlet b = Uint128::from(42u64);\nassert_eq!(b.u128(), 42);\n\nlet c = Uint128::from(70u32);\nassert_eq!(c.u128(), 70);\n```",
        "type": "integer",
        "precision": 128,
        "signed": false
      },
      {
        "name": "Coin",
        "type": "struct",
        "properties": {
          "amount": {
            "value": 1
          },
          "denom": {
            "value": 0
          }
        }
      },
      {
        "name": "alloc::vec::Vec<cosmwasm_std::coin::Coin>",
        "type": "array",
        "items": 2
      },
      {
        "name": "SudoMsg",
        "description": "SudoMsg is only exposed for internal Cosmos SDK modules to call.\nThis is showing how we can expose \"admin\" functionality than can not be called by\nexternal users or contracts, but only trusted (native/Go) code in the blockchain",
        "type": "enum",
        "cases": {
          "steal_funds": {
            "type": "named",
            "properties": {
              "amount": {
                "value": 3
              },
              "recipient": {
                "value": 0
              }
            }
          }
        }
      }
    ]
  },
  "responses": {
    "get_int": {
      "type": "v1",
      "root": 1,
      "definitions": [
        {
          "name": "u32",
          "type": "integer",
          "precision": 32,
          "signed": false
        },
        {
          "name": "IntResponse",
          "type": "struct",
          "properties": {
            "int": {
              "value": 0
            }
          }
        }
      ]
    },
    "other_balance": {
      "type": "v1",
      "root": 4,
      "definitions": [
        {
          "name": "String",
          "type": "string"
        },
        {
          "name": "Uint128",
          "description": "A thin wrapper around u128 that is using strings for JSON encoding/decoding,\nsuch that the full u128 range can be used for clients that convert JSON numbers to floats,\nlike JavaScript and jq.\n\n# Examples\n\nUse `from` to create instances of this and `u128` to get the value out:\n\n```\n# use cosmwasm_std::Uint128;\nlet a = Uint128::from(123u128);\nassert_eq!(a.u128(), 123);\n\nlet b = Uint128::from(42u64);\nassert_eq!(b.u128(), 42);\n\nlet c = Uint128::from(70u32);\nassert_eq!(c.u128(), 70);\n```",
          "type": "integer",
          "precision": 128,
          "signed": false
        },
        {
          "name": "Coin",
          "type": "struct",
          "properties": {
            "amount": {
              "value": 1
            },
            "denom": {
              "value": 0
            }
          }
        },
        {
          "name": "alloc::vec::Vec<cosmwasm_std::coin::Coin>",
          "type": "array",
          "items": 2
        },
        {
          "name": "AllBalanceResponse",
          "type": "struct",
          "properties": {
            "amount": {
              "description": "Returns all non-zero coins held by this account.",
              "value": 3
            }
          }
        }
      ]
    },
    "recurse": {
      "type": "v1",
      "root": 1,
      "definitions": [
        {
          "name": "Binary",
          "description": "Binary is a wrapper around Vec<u8> to add base64 de/serialization\nwith serde. It also adds some helper methods to help encode inline.\n\nThis is only needed as serde-json-{core,wasm} has a horrible encoding for Vec<u8>.\nSee also <https://github.com/CosmWasm/cosmwasm/blob/main/docs/MESSAGE_TYPES.md>.",
          "type": "binary"
        },
        {
          "name": "RecurseResponse",
          "type": "struct",
          "properties": {
            "hashed": {
              "description": "hashed is the result of running sha256 \"work+1\" times on the contract's human address",
              "value": 0
            }
          }
        }
      ]
    },
    "verifier": {
      "type": "v1",
      "root": 1,
      "definitions": [
        {
          "name": "String",
          "type": "string"
        },
        {
          "name": "VerifierResponse",
          "type": "struct",
          "properties": {
            "verifier": {
              "value": 0
            }
          }
        }
      ]
    }
  }
}
