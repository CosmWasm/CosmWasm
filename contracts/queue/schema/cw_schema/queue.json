{
  "contract_name": "queue",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "InstantiateMsg",
        "type": "struct",
        "properties": {}
      }
    ]
  },
  "execute": {
    "type": "v1",
    "root": 1,
    "definitions": [
      {
        "name": "i32",
        "type": "integer",
        "precision": 32,
        "signed": true
      },
      {
        "name": "ExecuteMsg",
        "type": "enum",
        "cases": {
          "dequeue": {
            "type": "named",
            "properties": {}
          },
          "enqueue": {
            "type": "named",
            "properties": {
              "value": {
                "value": 0
              }
            }
          }
        }
      }
    ]
  },
  "query": {
    "type": "v1",
    "root": 1,
    "definitions": [
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      },
      {
        "name": "QueryMsg",
        "type": "enum",
        "cases": {
          "count": {
            "type": "named",
            "properties": {}
          },
          "list": {
            "type": "named",
            "properties": {}
          },
          "open_iterators": {
            "description": "Opens the given number of iterators for no reason other than testing.\nReturns and `Empty` response.",
            "type": "named",
            "properties": {
              "count": {
                "value": 0
              }
            }
          },
          "reducer": {
            "type": "named",
            "properties": {}
          },
          "sum": {
            "type": "named",
            "properties": {}
          }
        }
      }
    ]
  },
  "migrate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "MigrateMsg",
        "type": "struct",
        "properties": {}
      }
    ]
  },
  "sudo": null,
  "responses": {
    "count": {
      "type": "v1",
      "root": 1,
      "definitions": [
        {
          "name": "u32",
          "type": "integer",
          "precision": 32,
          "signed": false
        },
        {
          "name": "CountResponse",
          "type": "struct",
          "properties": {
            "count": {
              "value": 0
            }
          }
        }
      ]
    },
    "list": {
      "type": "v1",
      "root": 2,
      "definitions": [
        {
          "name": "u32",
          "type": "integer",
          "precision": 32,
          "signed": false
        },
        {
          "name": "alloc::vec::Vec<u32>",
          "type": "array",
          "items": 0
        },
        {
          "name": "ListResponse",
          "type": "struct",
          "properties": {
            "early": {
              "description": "List all IDs lower than 0x20",
              "value": 1
            },
            "empty": {
              "description": "List an empty range, both bounded",
              "value": 1
            },
            "late": {
              "description": "List all IDs starting from 0x20",
              "value": 1
            }
          }
        }
      ]
    },
    "open_iterators": {
      "type": "v1",
      "root": 0,
      "definitions": [
        {
          "name": "Empty",
          "description": "An empty struct that serves as a placeholder in different places,\nsuch as contracts that don't set a custom message.\n\nIt is designed to be expressible in correct JSON and JSON Schema but\ncontains no meaningful data. Previously we used enums without cases,\nbut those cannot represented as valid JSON Schema (https://github.com/CosmWasm/cosmwasm/issues/451)",
          "type": "struct",
          "properties": {}
        }
      ]
    },
    "reducer": {
      "type": "v1",
      "root": 3,
      "definitions": [
        {
          "name": "i32",
          "type": "integer",
          "precision": 32,
          "signed": true
        },
        {
          "name": "(i32, i32)",
          "type": "tuple",
          "items": [
            0,
            0
          ]
        },
        {
          "name": "alloc::vec::Vec<(i32, i32)>",
          "type": "array",
          "items": 1
        },
        {
          "name": "ReducerResponse",
          "type": "struct",
          "properties": {
            "counters": {
              "value": 2
            }
          }
        }
      ]
    },
    "sum": {
      "type": "v1",
      "root": 1,
      "definitions": [
        {
          "name": "i32",
          "type": "integer",
          "precision": 32,
          "signed": true
        },
        {
          "name": "SumResponse",
          "type": "struct",
          "properties": {
            "sum": {
              "value": 0
            }
          }
        }
      ]
    }
  }
}
