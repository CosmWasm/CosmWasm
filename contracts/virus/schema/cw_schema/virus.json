{
  "contract_name": "virus",
  "contract_version": "0.0.0",
  "idl_version": "1.0.0",
  "instantiate": {
    "type": "v1",
    "root": 0,
    "definitions": [
      {
        "name": "InstantiateMsg",
        "type": "struct",
        "properties": {}
      }
    ]
  },
  "execute": {
    "type": "v1",
    "root": 2,
    "definitions": [
      {
        "name": "String",
        "type": "string"
      },
      {
        "name": "u32",
        "type": "integer",
        "precision": 32,
        "signed": false
      },
      {
        "name": "ExecuteMsg",
        "type": "enum",
        "cases": {
          "spread": {
            "type": "named",
            "properties": {
              "levels": {
                "description": "The number of levels of spreading. When set to 0, the contract performs a no-op.",
                "value": 1
              },
              "parent_path": {
                "description": "A slash separated path to the instance creating this one.\nThe root is the empty string.",
                "value": 0
              }
            }
          }
        }
      }
    ]
  },
  "query": null,
  "migrate": null,
  "sudo": null,
  "responses": null
}
