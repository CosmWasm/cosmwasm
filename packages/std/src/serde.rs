// This file simply re-exports some methods from serde_json
// The reason is two fold:
// 1. To easily ensure that all calling libraries use the same version (minimize code size)
// 2. To allow us to switch out to eg. serde-json-core more easily

use core::any::type_name;
use serde::{de::DeserializeOwned, Serialize};

use crate::Binary;
use crate::{StdError, StdResult};

/// Deserializes the given JSON bytes to a data structure.
///
/// Errors if the input is not valid JSON or cannot be deserialized to the given type.
pub fn from_json<T: DeserializeOwned>(value: impl AsRef<[u8]>) -> StdResult<T> {
    serde_json::from_slice(value.as_ref()).map_err(|e| StdError::parse_err(type_name::<T>(), e))
}

/// Serializes the given data structure as a JSON byte vector.
pub fn to_json_vec<T>(data: &T) -> StdResult<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    serde_json::to_vec(data).map_err(|e| StdError::serialize_err(type_name::<T>(), e))
}

/// Serializes the given data structure as a JSON string.
pub fn to_json_string<T>(data: &T) -> StdResult<String>
where
    T: Serialize + ?Sized,
{
    serde_json::to_string(data).map_err(|e| StdError::serialize_err(type_name::<T>(), e))
}

/// Serializes the given data structure as JSON bytes.
pub fn to_json_binary<T>(data: &T) -> StdResult<Binary>
where
    T: Serialize + ?Sized,
{
    to_json_vec(data).map(Binary::new)
}

#[cfg(test)]
mod tests {
    use super::*;
    use core::num::{
        NonZeroI128, NonZeroI16, NonZeroI32, NonZeroI64, NonZeroI8, NonZeroU128, NonZeroU16,
        NonZeroU32, NonZeroU64, NonZeroU8,
    };
    use proptest::{prop_assert_eq, property_test};
    use serde::Deserialize;

    use crate::msgpack::{from_msgpack, to_msgpack_vec};

    #[derive(Serialize, Deserialize, Debug, PartialEq)]
    #[serde(rename_all = "snake_case")]
    enum SomeMsg {
        Refund {},
        ReleaseAll {
            image: String,
            amount: u32,
            time: u64,
            karma: i32,
        },
        Cowsay {
            text: String,
        },
    }

    #[test]
    fn to_json_vec_works() {
        let msg = SomeMsg::Refund {};
        let serialized = to_json_vec(&msg).unwrap();
        assert_eq!(serialized, br#"{"refund":{}}"#);

        let msg = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 9007199254740999, // Number.MAX_SAFE_INTEGER + 7
            karma: -17,
        };
        let serialized = String::from_utf8(to_json_vec(&msg).unwrap()).unwrap();
        assert_eq!(
            serialized,
            r#"{"release_all":{"image":"foo","amount":42,"time":9007199254740999,"karma":-17}}"#
        );
    }

    #[test]
    fn from_json_works() {
        let deserialized: SomeMsg = from_json(br#"{"refund":{}}"#).unwrap();
        assert_eq!(deserialized, SomeMsg::Refund {});

        let expected = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 18446744073709551615,
            karma: -17,
        };
        // &[u8]
        let deserialized: SomeMsg = from_json(
            br#"{"release_all":{"image":"foo","amount":42,"time":18446744073709551615,"karma":-17}}"#,
        )
        .unwrap();
        assert_eq!(deserialized, expected);

        // &str
        let deserialized: SomeMsg = from_json(
            r#"{"release_all":{"image":"foo","amount":42,"time":18446744073709551615,"karma":-17}}"#,
        )
        .unwrap();
        assert_eq!(deserialized, expected);
    }

    #[test]
    fn from_json_or_binary() {
        let msg = SomeMsg::Refund {};
        let serialized: Binary = to_json_binary(&msg).unwrap();

        let parse_binary: SomeMsg = from_json(&serialized).unwrap();
        assert_eq!(parse_binary, msg);

        let parse_slice: SomeMsg = from_json(serialized.as_slice()).unwrap();
        assert_eq!(parse_slice, msg);
    }

    #[test]
    fn to_json_vec_works_for_special_chars() {
        let msg = SomeMsg::Cowsay {
            text: "foo\"bar\\\"bla".to_string(),
        };
        let serialized = String::from_utf8(to_json_vec(&msg).unwrap()).unwrap();
        assert_eq!(serialized, r#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#);
    }

    #[test]
    fn from_json_works_for_special_chars() {
        let deserialized: SomeMsg = from_json(br#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#).unwrap();
        assert_eq!(
            deserialized,
            SomeMsg::Cowsay {
                text: "foo\"bar\\\"bla".to_string(),
            }
        );
    }

    #[test]
    fn to_json_string_works() {
        let msg = SomeMsg::Refund {};
        let serialized = to_json_string(&msg).unwrap();
        assert_eq!(serialized, r#"{"refund":{}}"#);

        let msg = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 9007199254740999, // Number.MAX_SAFE_INTEGER + 7
            karma: -17,
        };
        let serialized = to_json_string(&msg).unwrap();
        assert_eq!(
            serialized,
            r#"{"release_all":{"image":"foo","amount":42,"time":9007199254740999,"karma":-17}}"#
        );
    }

    macro_rules! test_integer {
        ($($ty:ty),+$(,)?) => {
            $(
                ::paste::paste! {
                    #[property_test]
                    fn [<test_ $ty:snake:lower _encoding>](input: $ty) {
                        let primitive = input.get();

                        // Verify that the serialization is the same as the primitive
                        let serialized = to_json_string(&input).unwrap();
                        let serialized_primitive = to_json_string(&primitive).unwrap();
                        prop_assert_eq!(serialized.as_str(), serialized_primitive.as_str());

                        // Verify that the serialized primitive can be deserialized
                        let deserialized: $ty = from_json(serialized_primitive).unwrap();
                        assert_eq!(deserialized, input);

                        // Verify that zero is not allowed
                        assert!(from_json::<$ty>("0").is_err());

                        // Verify that the msgpack encoding is the same as the primitive
                        let serialized = to_msgpack_vec(&input).unwrap();
                        let serialized_primitive = to_msgpack_vec(&primitive).unwrap();
                        prop_assert_eq!(serialized.as_slice(), serialized_primitive.as_slice());

                        // Verify that the serialized primitive can be deserialized
                        let deserialized: $ty = from_msgpack(&serialized_primitive).unwrap();
                        prop_assert_eq!(deserialized, input);
                    }
                }
            )+
        };
    }

    test_integer! {
        NonZeroU8,
        NonZeroU16,
        NonZeroU32,
        NonZeroU64,
        NonZeroU128,
    }

    test_integer! {
        NonZeroI8,
        NonZeroI16,
        NonZeroI32,
        NonZeroI64,
        NonZeroI128,
    }
}
