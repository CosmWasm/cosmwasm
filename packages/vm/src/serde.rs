//! This file simply re-exports some methods from serde_json
//! The reason is two fold:
//! 1. To easily ensure that all calling libraries use the same version (minimize code size)
//! 2. To allow us to switch out to eg. serde-json-core more easily
use serde::{Deserialize, Serialize};
use std::any::type_name;

use crate::errors::{VmError, VmResult};
use crate::security_limits::{
    enforce_deserialization_limit, MAX_DESERIALIZATION_BYTES, MAX_DESERIALIZATION_DEPTH,
};

/// Deserializes JSON data into a document of type `T`.
///
/// The deserialization limit ensure it is not possible to slow down the execution by
/// providing overly large JSON documents.
pub fn from_slice<'a, T>(value: &'a [u8], deserialization_limit: usize) -> VmResult<T>
where
    T: Deserialize<'a>,
{
    // Enforce a global maximum regardless of what limit is passed in
    let enforced_limit = enforce_deserialization_limit(value.len(), deserialization_limit);

    if value.len() > enforced_limit {
        return Err(VmError::deserialization_limit_exceeded(
            value.len(),
            enforced_limit,
        ));
    }

    // Create a depth-limiting deserializer
    let mut deserializer = serde_json::Deserializer::from_slice(value);
    deserializer.disable_recursion_limit();
    let deserializer = serde_json::Deserializer::with_recursion_limit(
        deserializer,
        MAX_DESERIALIZATION_DEPTH as usize,
    );

    // Use the depth-limited deserializer
    let result = T::deserialize(deserializer);
    result.map_err(|e| VmError::parse_err(type_name::<T>(), e))
}

pub fn to_vec<T>(data: &T) -> VmResult<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    serde_json::to_vec(data).map_err(|e| VmError::serialize_err(type_name::<T>(), e))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;

    const LIMIT: usize = 20_000;

    #[derive(Serialize, Deserialize, Debug, PartialEq)]
    #[serde(rename_all = "snake_case")]
    enum SomeMsg {
        Refund {},
        ReleaseAll {
            image: String,
            amount: u32,
            time: u64,
            karma: i32,
        },
        Cowsay {
            text: String,
        },
    }

    #[test]
    fn from_slice_works() {
        let deserialized: SomeMsg = from_slice(br#"{"refund":{}}"#, LIMIT).unwrap();
        assert_eq!(deserialized, SomeMsg::Refund {});

        let deserialized: SomeMsg = from_slice(
            br#"{"release_all":{"image":"foo","amount":42,"time":18446744073709551615,"karma":-17}}"#, LIMIT
        )
        .unwrap();
        assert_eq!(
            deserialized,
            SomeMsg::ReleaseAll {
                image: "foo".to_string(),
                amount: 42,
                time: 18446744073709551615,
                karma: -17
            }
        );
    }

    #[test]
    fn from_slice_works_for_special_chars() {
        let deserialized: SomeMsg =
            from_slice(br#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#, LIMIT).unwrap();
        assert_eq!(
            deserialized,
            SomeMsg::Cowsay {
                text: "foo\"bar\\\"bla".to_string(),
            }
        );
    }

    #[test]
    fn from_slice_errors_when_exceeding_deserialization_limit() {
        let result = from_slice::<SomeMsg>(br#"{"refund":{}}"#, 5);
        match result.unwrap_err() {
            VmError::DeserializationLimitExceeded {
                length, max_length, ..
            } => {
                assert_eq!(length, 13);
                assert_eq!(max_length, 5);
            }
            err => panic!("Unexpected error: {err:?}"),
        }
    }

    #[test]
    fn from_slice_enforces_global_limit() {
        // Test that even if a large limit is provided, it's capped by MAX_DESERIALIZATION_BYTES
        let very_large_limit = MAX_DESERIALIZATION_BYTES * 2;
        let data = vec![b'{'; MAX_DESERIALIZATION_BYTES + 1000]; // Data slightly larger than the max
        let result = from_slice::<serde_json::Value>(&data, very_large_limit);
        match result.unwrap_err() {
            VmError::DeserializationLimitExceeded {
                length, max_length, ..
            } => {
                assert_eq!(length, MAX_DESERIALIZATION_BYTES + 1000);
                assert_eq!(max_length, MAX_DESERIALIZATION_BYTES);
            }
            err => panic!("Unexpected error: {err:?}"),
        }
    }

    #[test]
    fn from_slice_prevents_deeply_nested_json() {
        // Create a deeply nested JSON structure beyond our recursion limit
        let mut json = String::from("{\"data\":");
        for _ in 0..MAX_DESERIALIZATION_DEPTH as usize + 5 {
            json.push_str("{\"data\":");
        }
        json.push_str("\"value\"");
        for _ in 0..MAX_DESERIALIZATION_DEPTH as usize + 5 {
            json.push('}');
        }

        // This should fail due to recursion depth
        let result = from_slice::<serde_json::Value>(json.as_bytes(), LIMIT);
        assert!(result.is_err());
        let err = result.unwrap_err();
        match err {
            VmError::ParseErr { .. } => {
                // We expect a parsing error related to recursion depth
                assert!(
                    err.to_string().contains("recursion")
                        || err.to_string().contains("depth")
                        || err.to_string().contains("nested")
                );
            }
            err => panic!("Unexpected error: {err:?}"),
        }
    }

    #[test]
    fn to_vec_works() {
        let msg = SomeMsg::Refund {};
        let serialized = to_vec(&msg).unwrap();
        assert_eq!(serialized, br#"{"refund":{}}"#);

        let msg = SomeMsg::ReleaseAll {
            image: "foo".to_string(),
            amount: 42,
            time: 9007199254740999, // Number.MAX_SAFE_INTEGER + 7
            karma: -17,
        };
        let serialized = String::from_utf8(to_vec(&msg).unwrap()).unwrap();
        assert_eq!(
            serialized,
            r#"{"release_all":{"image":"foo","amount":42,"time":9007199254740999,"karma":-17}}"#
        );
    }

    #[test]
    fn to_vec_works_for_special_chars() {
        let msg = SomeMsg::Cowsay {
            text: "foo\"bar\\\"bla".to_string(),
        };
        let serialized = String::from_utf8(to_vec(&msg).unwrap()).unwrap();
        assert_eq!(serialized, r#"{"cowsay":{"text":"foo\"bar\\\"bla"}}"#);
    }
}
