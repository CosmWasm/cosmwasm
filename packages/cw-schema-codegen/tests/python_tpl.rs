use cw_schema::Schemaifier;
use serde::{Deserialize, Serialize};
use std::io::Write;

/// This is a struct level documentation for enum type
#[derive(Schemaifier, Serialize, Deserialize, PartialEq, Debug)]
pub enum SomeEnum {
    /// Field1 docs
    Field1,

    /// Field2 docs
    Field2(u32, u32),

    /// Field3 docs
    Field3 {
        /// `a` field docs
        a: String,

        /// `b` field docs
        b: u32,
    },
}

/// This is a struct level documentation for unit struct
#[derive(Schemaifier, Serialize, Deserialize, PartialEq, Debug)]
pub struct UnitStructure;

/// This is a struct level documentation for tuple
#[derive(Schemaifier, Serialize, Deserialize, PartialEq, Debug)]
pub struct TupleStructure(u32, String, u128);

/// This is a struct level documentation for named structure
#[derive(Schemaifier, Serialize, Deserialize, PartialEq, Debug)]
pub struct NamedStructure {
    /// `a` field docs
    a: String,

    /// `b` field docs
    b: u8,

    /// `c` field docs
    c: SomeEnum,
}

#[derive(Schemaifier, Serialize, Deserialize, PartialEq, Debug)]
pub struct AllSimpleTypesAndDocs {
    array_field: Vec<String>,
    float_field: f32,
    double_field: f64,
    bool_field: bool,
    string_field: String,
    int_field: i64,
    bytes_field: cosmwasm_std::Binary,
    opt_field: Option<String>,
    byte_field: u8,
    decimal_field: cosmwasm_std::Decimal,
    address_field: cosmwasm_std::Addr,
    checksum_field: cosmwasm_std::Checksum,
    hexbinary_field: cosmwasm_std::HexBinary,
    timestamp_field: cosmwasm_std::Timestamp,
    unit_field: (),
}

#[test]
fn simple_enum() {
    // generate the schemas for each of the above types
    let schemas = [
        cw_schema::schema_of::<SomeEnum>(),
        cw_schema::schema_of::<UnitStructure>(),
        cw_schema::schema_of::<TupleStructure>(),
        cw_schema::schema_of::<NamedStructure>(),
    ];

    // run the codegen to typescript
    for schema in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            panic!();
        };

        let output = schema
            .definitions
            .iter()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::python::process_node(&mut buf, &schema, node).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        insta::assert_snapshot!(output);
    }
}

macro_rules! validator {
    ($ty:ty, $example:expr) => {{
        (
            ::std::any::type_name::<$ty>().replace("::", "_"),
            cw_schema::schema_of::<$ty>(),
            serde_json::to_string(&$example).unwrap(),
            {
                let a: Box<dyn FnOnce(&str)> = Box::new(|output| {
                    let result = serde_json::from_str::<$ty>(output).unwrap();
                    assert_eq!(result, $example);
                });
                a
            },
        )
    }};
}

#[test]
fn assert_validity() {
    let schemas = [
        validator!(SomeEnum, SomeEnum::Field1),
        validator!(SomeEnum, SomeEnum::Field2(10, 23)),
        validator!(
            SomeEnum,
            SomeEnum::Field3 {
                a: "sdf".to_string(),
                b: 12,
            }
        ),
        validator!(UnitStructure, UnitStructure {}),
        validator!(TupleStructure, TupleStructure(10, "aasdf".to_string(), 2)),
        validator!(
            NamedStructure,
            NamedStructure {
                a: "awer".to_string(),
                b: 4,
                c: SomeEnum::Field1,
            }
        ),
        validator!(
            AllSimpleTypesAndDocs,
            AllSimpleTypesAndDocs {
                array_field: vec!["abc".to_string(), "def".to_string()],
                float_field: 10.2,
                double_field: 10.232323,
                bool_field: true,
                string_field: "sdfsdf".to_string(),
                int_field: -10,
                bytes_field: cosmwasm_std::Binary::new(vec![0x1, 0x2, 0x3]),
                opt_field: Some("sdfsdfwer".to_string()),
                byte_field: 9,
                decimal_field: cosmwasm_std::Decimal::one(),
                address_field: cosmwasm_std::Addr::unchecked("some_address"),
                checksum_field: cosmwasm_std::Checksum::generate(&[0x10]),
                hexbinary_field: cosmwasm_std::HexBinary::from_hex("FAFAFA").unwrap(),
                timestamp_field: cosmwasm_std::Timestamp::from_seconds(100),
                unit_field: (),
            }
        ),
    ];

    for (type_name, schema, example, validator) in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            unreachable!();
        };

        let schema_output = schema
            .definitions
            .iter()
            .rev()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::python::process_node(&mut buf, &schema, node).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        let mut file = tempfile::NamedTempFile::with_suffix(".py").unwrap();
        file.write_all(schema_output.as_bytes()).unwrap();
        file.write_all(
            format!(
                "import sys; print({type_name}.model_validate_json('{example}').model_dump_json())"
            )
            .as_bytes(),
        )
        .unwrap();
        file.flush().unwrap();

        let output = std::process::Command::new("python3")
            .arg(file.path())
            .output()
            .unwrap();

        assert!(
            output.status.success(),
            "stdout: {stdout}, stderr: {stderr}\n\n schema:\n {schema_output}",
            stdout = String::from_utf8_lossy(&output.stdout),
            stderr = String::from_utf8_lossy(&output.stderr),
        );

        validator(&String::from_utf8_lossy(&output.stdout))
    }
}
