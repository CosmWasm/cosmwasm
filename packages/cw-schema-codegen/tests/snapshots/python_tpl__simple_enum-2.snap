---
source: packages/cw-schema-codegen/tests/python_tpl.rs
expression: output
---
# This code is @generated by cw-schema-codegen. Do not modify this manually.

import typing
import decimal
from pydantic import BaseModel, RootModel



class python_tpl_UnitStructure(RootModel):
    """
    This is a struct level documentation for unit struct
    """
    root: None
